File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob_the_money(self, ind: int, nums: List[int],dp: List[int]):
        if ind == len(nums)-1:
            return nums[ind]
        if ind >= len(nums):
            return 0
        if dp[ind] != -1:
            return dp[ind]
        not_take = self.rob_the_money(ind+1,nums,dp)
        take = nums[ind] + self.rob_the_money(ind+2,nums,dp)
        dp[ind] = max(take,not_take)
        return dp[ind]
    def rob(self, nums: List[int]) -> int:
        n=len(nums)
        if n==1:
            return nums[0]
        dp = [-1]*n
        return self.rob_the_money(0,nums,dp)
        

File: 0198-house-robber/0198-house-robber.cpp
class Solution {
private:
    int dp[105];
    int robIt(int ind, vector<int>& nums, int n)
    {
        if(ind >= n) return 0;
        if(dp[ind]!=-1) return dp[ind];
        // 2 cases broke current house and move to cur+2 house or not broke it
        int notTake = 0 + robIt(ind+1,nums,n);
        int take = nums[ind] + robIt(ind+2,nums,n);

        return dp[ind] = max(take,notTake);

    }
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        memset(dp,-1,sizeof(dp));
        return robIt(0,nums,n);
    }
};